src/HRayData.hs

```diff
@@ -0,0 +1,131 @@
+{-# LANGUAGE AllowAmbiguousTypes       #-}
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE RecordWildCards           #-}
+{-# LANGUAGE ScopedTypeVariables       #-}
+
+module HRayData
+    ( AABB
+    , AnyHit
+    , BoundingVolume
+    , ClosestHit
+    , Color
+    , Context
+    , D2
+    , D3
+    , Intersect
+    , Intersection
+    , Ray
+    , RayResult
+    , Scene
+    , SceneObject
+    -- | Functions
+    , getDouble
+    , getVector2
+    , getVector3
+    , raycast
+    ) where
+
+import Data.Map (Map, (!))
+import Linear.V2 (V2)
+import Linear.V3 (V3)
+
+import qualified Data.Map as Map hiding (Map)
+import qualified Linear.V2 as V2 hiding (V2)
+import qualified Linear.V3 as V3 hiding (V3)
+
+newtype D2 =
+    D2 (V2 Double)
+
+newtype D3 =
+    D3 (V3 Double)
+
+newtype Color =
+    Color D3
+
+data Ray = Ray
+    { origin    :: D3
+    , direction :: D3
+    }
+
+newtype Intersection =
+    Intersection D3
+
+-- | Ray Result is either a Color
+-- or a ray generator and a function to convert the ray results into Color
+data RayResult
+    = RayResult Color
+    | RayResultAssembly [Ray]
+                        ([RayResult] -> Color)
+
+newtype Intersect =
+    Intersect (Context -> Maybe Intersection)
+
+-- | Called when the intersection point was the closest.
+newtype ClosestHit =
+    ClosestHit RayResult
+
+-- | Called when looking for any intersection point.
+-- Returns a RayResult to stop loooking for intersection points or Nothing to keep looking.
+newtype AnyHit =
+    AnyHit (Maybe RayResult)
+
+data Context = Context
+    { v1s :: Map String Double
+    , v2s :: Map String D2
+    , v3s :: Map String D3
+    }
+
+-- | Is used to determine if an underlying SceneObject has a chance of intersecting with the ray.
+class BoundingVolume b where
+    intersect :: Intersect
+
+-- | Axis Alligned Bounding Box.
+data AABB = AABB
+    { min :: D3
+    , max :: D3
+    }
+
+-- | TODO: implement intesection with the bounding box
+instance BoundingVolume AABB where
+    intersect = undefined
+
+-- | Scene object is fully defined by it's behaviour and context.
+data SceneObject =
+    SceneObject Context
+                Intersect
+                ClosestHit
+                AnyHit
+                AABB -- | TODO: parametrize with a bounding volume type.
```

Comment:
To do this you can add existensial data type like this:

```haskell
data AnyBoundingVolume = forall b . BoundingVolume b => Any b
```

But this design looks really strange to me. Why have function

```haskell
interesect :: BoundingVolume b => Intersect
```
instead of just storing `Intersect` data type inside `SceneObject`. But I see that `Intersect` is already stored here. Which makes kinda strange the need to store type `AABB` itself...

Reply:
Okay, I think I understood this.